bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v2/product/ProductImpl.java

```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+ *
+ *    Copyright 2021 Adobe. All rights reserved.
+ *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License. You may obtain a copy
+ *    of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software distributed under
+ *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ *    OF ANY KIND, either express or implied. See the License for the specific language
+ *    governing permissions and limitations under the License.
+ *
+ ******************************************************************************/
+
+package com.adobe.cq.commerce.core.components.internal.models.v2.product;
+
+import javax.annotation.PostConstruct;
+
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.models.annotations.Model;
+
+import com.adobe.cq.commerce.core.components.models.product.Product;
+import com.adobe.cq.commerce.magento.graphql.ConfigurableProduct;
+import com.adobe.cq.commerce.magento.graphql.GroupedProduct;
+import com.adobe.cq.commerce.magento.graphql.ProductInterface;
+
+@Model(
+    adaptables = SlingHttpServletRequest.class,
+    adapters = Product.class,
+    resourceType = ProductImpl.RESOURCE_TYPE,
+    cache = true)
+public class ProductImpl extends com.adobe.cq.commerce.core.components.internal.models.v1.product.ProductImpl implements Product {
+
+    public static final String RESOURCE_TYPE = "core/cif/components/commerce/product/v2/product";
+
+    @PostConstruct
+    protected void initModel() {
+        super.initModel();
+        if (productRetriever != null) {
+            productRetriever.extendProductQueryWith(p -> p.staged());
```

Comment:
In case someone wants to customize the component via Sling model delegation and uses the `extendProductQueryWith` or `extendVariantQueryWith` methods and does not include the `staged` attribute, the staging functionality will no longer work. The way these methods work is that the last call wins. Any customizations done before will no longer be respected.

So I would rather:
* either change the way `extendProductQueryWith` and `extendVariantQueryWith` work so that you can call it multiple times and customizations are "merged".
* or create a separate `ProductRetriever` implementation (based on `AbstractProductRetriever`) and add the staged attribute there. I prefer this way, since we will need much more attributes for gift card support.

Reply:
Good point, I'll fix that.

Reply:
I finally went for option 1) see the change in `AbstractProductRetriever`